[RISCV] Support for Zvabd fast-track proposal

wangpc-pp opened the pull request:
**NOTE: This is an early-access support only for evaluation, not going to be merged and may be changable during the process of fast-track proposal.**

----

This PR contains the MC/CodeGen support of ByteDance's proposed Zvabd (Vector Absolute Difference) extension.

Support of these instructions are added:

- Vector Single-Width Signed/Unsigned Integer Absolute Difference
- Vector Widening Signed/Unsigned Integer Absolute Difference and
  Accumulate

Fast-track Progress: https://lf-riscv.atlassian.net/wiki/spaces/VXXX/pages/166690866/Fast-Track+Proposal+for+integer+vector+absolute+difference+instructions
Mailing List: https://lists.riscv.org/g/sig-vector/topic/110210092
Github Repo: https://github.com/riscv/integer-vector-absolute-difference

wangpc-pp committed:
[RISCV][MC] Support experimental Zvabd instructions

The `Zvabd` is for `RISC-V Integer Vector Absolute Difference` and
it provides 5 instructions:

* `vabs.v`: Vector Signed Integer Absolute.
* `vabd.vv`: Vector Signed Integer Absolute Difference.
* `vabdu.vv`: Vector Unsigned Integer Absolute Difference.
* `vwabda.vv`: Vector Signed Integer Absolute Difference And Accumulate.
* `vwabdau.vv`: Vector Unsigned Integer Absolute Difference And Accumulate.

Doc: https://github.com/riscv/integer-vector-absolute-difference

Pull Request: https://github.com/llvm/llvm-project/pull/180139

wangpc-pp committed:
[RISCV][CodeGen] Lowering abds/abdu to Zvabd instructions

We directly lower `ISD::ABDS`/`ISD::ABDU` to `Zvabd` instructions.

Note that we only support SEW=8/16 for `vabd.vv`/`vabdu.vv`.

Pull Request: https://github.com/llvm/llvm-project/pull/180141

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -29,6 +29,7 @@
 #include "llvm/CodeGen/MachineJumpTableInfo.h"
 #include "llvm/CodeGen/MachineRegisterInfo.h"
 #include "llvm/CodeGen/SelectionDAGAddressAnalysis.h"
+#include "llvm/CodeGen/SelectionDAGNodes.h"
```
what's special about this patch that requires this include?

wangpc-pp replied:
It's an unnecessary include caused by auto-includer, I will remove it.

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoV.td:
```diff
@@ -575,15 +575,16 @@ multiclass VALU_IV_X<string opcodestr, bits<6> funct6> {
 }
 
-multiclass VALU_IV_V_X_I<string opcodestr, bits<6> funct6>
+multiclass VALU_IV_V_X_I<string opcodestr, bits<6> funct6,
```
These classes don't appear to be referenced elsewhere in the patch. Why do they need to be chaned?

wangpc-pp replied:
The initial proposal contained `.vi` variants but we deleted them later. I should remove this change here. :-)

wangpc-pp committed:
[RISCV][CodeGen] Lower `ISD::ABS` to Zvabd instructions

We add pseudos/patterns for `vabs.v` instruction and handle the
lowering in `RISCVTargetLowering::lowerABS`.

Pull Request: https://github.com/llvm/llvm-project/pull/180142

wangpc-pp committed:
[RISCV][CodeGen] Combine vwaddu+vabd(u) to vwabdacc(u)

Note that we only support SEW=8/16 for `vwabdacc(u)`.

Pull Request: https://github.com/llvm/llvm-project/pull/180162

topperc commented on llvm/test/CodeGen/RISCV/rvv/fixed-vectors-sad.ll:
```diff
@@ -135,6 +183,43 @@ define signext i32 @sad_2block_16xi8_as_i32(ptr %a, ptr %b, i32 signext %stridea
+; ZVABD-NEXT:    vsetvli zero, zero, e16, m2, ta, ma
+; ZVABD-NEXT:    vzext.vf2 v12, v10
+; ZVABD-NEXT:    vsetvli zero, zero, e8, m1, ta, ma
+; ZVABD-NEXT:    vwabdaccu.vv v12, v8, v9
```
Is there a codegen test for vwabdacc.vv?

wangpc-pp replied:
I will try to add one.

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -20594,6 +20645,8 @@ SDValue RISCVTargetLowering::PerformDAGCombine(SDNode *N,
     if (SDValue V = combineVqdotAccum(N, DAG, Subtarget))
       return V;
     return combineToVWMACC(N, DAG, Subtarget);
+  case RISCVISD::VWADDU_VL:
```
What about VWADDU_W_VL?

wangpc-pp committed:
[RISCV][TTI] Adjust the cost of `llvm.abs` intrinsic when `Zvabd` exists

When `Zvabd` exists, `llvm.abs` is lowered to `vabs.v` so the cost
is 1.

Pull Request: https://github.com/llvm/llvm-project/pull/180146

wangpc-pp committed:
[RISCV] Add precommit test for vwaddu_wv+vabd(u) to vwabdacc(u) combine

wangpc-pp committed:
[RISCV] Combine vwaddu_wv+vabd(u) to vwabdacc(u)

Note that we only support SEW=8/16 for `vwabdacc(u)`.

wangpc-pp committed:
[RISCV] Change Commutable to 0 for Zvabd

Commit:
[Clang][RISCV] Add Zvabd intrinsics

wangpc-pp commented:
Thanks for reviewing this PR! I will split it into several PRs and commit them separately.
For the record (updating):
1. https://github.com/llvm/llvm-project/pull/180139

wangpc-pp committed:
[RISCV] Add precommit test for vwabdacc(u) combine

Pull Request: https://github.com/llvm/llvm-project/pull/180161
